Race condition in task replacement causes duplicate execution

## Problem

The `test_rescheduling_later` test is **intermittently failing** in CI because tasks are being executed twice when using `docket.replace()`. The issue occurs due to a race condition between task scheduling and replacement:

1. Task is scheduled for `now() + 10ms` via `docket.add()`
2. `docket.replace()` is called immediately after to reschedule for `now() + 100ms`  
3. The worker's scheduler loop (running every 250ms) moves the first task from queue to stream before `replace` can cancel it
4. Both the original task (in stream) and replacement task (in queue/stream) get executed

**Example failure:** https://github.com/chrisguidry/docket/actions/runs/16608341559/attempts/1?pr=148

## Root Cause

The `_cancel()` method in `src/docket/docket.py:474` only removes tasks from:
- Known task key (hash)
- Parked task key (hash) 
- Queue (sorted set)

It does **not** remove tasks that have already been moved to the stream by the scheduler. Once a task is in the stream, `replace()` cannot cancel it, leading to duplicate execution.

## Error Details

```
FAILED tests/test_fundamentals.py::test_rescheduling_later[asyncio] - AssertionError: Expected mock to have been awaited once. Awaited 2 times.
```

The debug logs show the scheduler actively moving tasks multiple times, confirming the race condition.

## Proposed Solutions

### Option 1: Cancellation Tracking (Recommended)

Add a Redis set to track cancelled task keys that workers check before execution:

**Implementation:**
1. Add `cancelled_tasks` set to Redis data model
2. When `replace()` is called, add the old task key to this set  
3. Workers check this set before executing tasks and skip cancelled ones
4. Implement periodic cleanup of old cancelled task entries

**Pros:**
- No performance impact on streams
- Simple to implement and understand
- Works regardless of where the task currently resides
- Minimal Redis operations

**Cons:** 
- Requires additional Redis storage
- Needs cleanup mechanism for cancelled task keys

### Option 3: Stream Message ID Tracking

Track Redis stream message IDs to enable direct deletion:

**Implementation:**
1. Modify stream scheduling to capture and store message IDs
2. Store message ID in the `known_task_key` when task moves to stream
3. Update `_cancel()` to delete stream messages by ID when available

**Pros:**
- Direct deletion without scanning
- Clean architectural approach
- No additional data structures needed

**Cons:**
- Requires changes to scheduler Lua script to return message IDs
- More complex implementation across multiple components
- Need to handle cases where message ID isn't available

## Test Case

The failing test:
```python
async def test_rescheduling_later(
    docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
):
    key = f"my-cool-task:{uuid4()}"
    
    soon = now() + timedelta(milliseconds=10)  # Very short delay
    await docket.add(the_task, soon, key=key)("a", "b", c="c")
    
    later = now() + timedelta(milliseconds=100)
    await docket.replace(the_task, later, key=key)("b", "c", c="d")  # Should cancel first task
    
    await worker.run_until_finished()
    
    the_task.assert_awaited_once_with("b", "c", c="d")  # Fails - called twice
```

## Files Affected

- `src/docket/docket.py:474` - `_cancel()` method needs enhancement
- `src/docket/worker.py` - May need changes for message ID tracking (Option 3)
- `tests/test_fundamentals.py:test_rescheduling_later` - Currently failing

## Priority

High - This affects the core functionality of task replacement and causes intermittent test failures in CI.

--- a/src/docket/docket.py
+++ b/src/docket/docket.py
@@ -16,6 +16,7 @@ from typing import (
     Mapping,
     NoReturn,
     ParamSpec,
+    Protocol,
     Self,
     Sequence,
     TypedDict,
@@ -27,7 +28,6 @@ from typing import (
 import redis.exceptions
 from opentelemetry import propagate, trace
 from redis.asyncio import ConnectionPool, Redis
-from redis.asyncio.client import Pipeline
 from uuid_extensions import uuid7
 
 from .execution import (
@@ -53,6 +53,18 @@ from .instrumentation import (
 
 logger: logging.Logger = logging.getLogger(__name__)
 tracer: trace.Tracer = trace.get_tracer(__name__)
+
+
+class _schedule_task(Protocol):
+    async def __call__(
+        self, keys: list[str], args: list[str | float | bytes]
+    ) -> str: ...  # pragma: no cover
+
+
+class _cancel_task(Protocol):
+    async def __call__(
+        self, keys: list[str], args: list[str]
+    ) -> str: ...  # pragma: no cover
 
 
 P = ParamSpec("P")
@@ -131,6 +143,8 @@ class Docket:
 
     _monitor_strikes_task: asyncio.Task[None]
     _connection_pool: ConnectionPool
+    _schedule_task_script: _schedule_task | None
+    _cancel_task_script: _cancel_task | None
 
     def __init__(
         self,
@@ -156,6 +170,8 @@ class Docket:
         self.url = url
         self.heartbeat_interval = heartbeat_interval
         self.missed_heartbeats = missed_heartbeats
+        self._schedule_task_script = None
+        self._cancel_task_script = None
 
     @property
     def worker_group_name(self) -> str:
@@ -300,9 +316,7 @@ class Docket:
             execution = Execution(function, args, kwargs, when, key, attempt=1)
 
             async with self.redis() as redis:
-                async with redis.pipeline() as pipeline:
-                    await self._schedule(redis, pipeline, execution, replace=False)
-                    await pipeline.execute()
+                await self._schedule(redis, execution, replace=False)
 
             TASKS_ADDED.add(1, {**self.labels(), **execution.general_labels()})
             TASKS_SCHEDULED.add(1, {**self.labels(), **execution.general_labels()})
@@ -361,9 +375,7 @@ class Docket:
             execution = Execution(function, args, kwargs, when, key, attempt=1)
 
             async with self.redis() as redis:
-                async with redis.pipeline() as pipeline:
-                    await self._schedule(redis, pipeline, execution, replace=True)
-                    await pipeline.execute()
+                await self._schedule(redis, execution, replace=True)
 
             TASKS_REPLACED.add(1, {**self.labels(), **execution.general_labels()})
             TASKS_CANCELLED.add(1, {**self.labels(), **execution.general_labels()})
@@ -383,9 +395,7 @@ class Docket:
             },
         ):
             async with self.redis() as redis:
-                async with redis.pipeline() as pipeline:
-                    await self._schedule(redis, pipeline, execution, replace=False)
-                    await pipeline.execute()
+                await self._schedule(redis, execution, replace=False)
 
         TASKS_SCHEDULED.add(1, {**self.labels(), **execution.general_labels()})
 
@@ -400,9 +410,7 @@ class Docket:
             attributes={**self.labels(), "docket.key": key},
         ):
             async with self.redis() as redis:
-                async with redis.pipeline() as pipeline:
-                    await self._cancel(pipeline, key)
-                    await pipeline.execute()
+                await self._cancel(redis, key)
 
         TASKS_CANCELLED.add(1, self.labels())
 
@@ -423,10 +431,17 @@ class Docket:
     async def _schedule(
         self,
         redis: Redis,
-        pipeline: Pipeline,
         execution: Execution,
         replace: bool = False,
     ) -> None:
+        """Schedule a task atomically.
+
+        Handles:
+        - Checking for task existence
+        - Cancelling existing tasks when replacing
+        - Adding tasks to stream (immediate) or queue (future)
+        - Tracking stream message IDs for later cancellation
+        """
         if self.strike_list.is_stricken(execution):
             logger.warning(
                 "%r is stricken, skipping schedule of %r",
@@ -449,32 +464,133 @@ class Docket:
         key = execution.key
         when = execution.when
         known_task_key = self.known_task_key(key)
+        is_immediate = when <= datetime.now(timezone.utc)
 
+        # Lock per task key to prevent race conditions between concurrent operations
         async with redis.lock(f"{known_task_key}:lock", timeout=10):
-            if replace:
-                await self._cancel(pipeline, key)
-            else:
-                # if the task is already in the queue or stream, retain it
-                if await redis.exists(known_task_key):
-                    logger.debug(
-                        "Task %r is already in the queue or stream, not scheduling",
-                        key,
-                        extra=self.labels(),
-                    )
-                    return
+            if self._schedule_task_script is None:
+                self._schedule_task_script = cast(
+                    _schedule_task,
+                    redis.register_script(
+                        # KEYS: stream_key, known_key, parked_key, queue_key
+                        # ARGV: task_key, when_timestamp, is_immediate, replace, ...message_fields
+                        """
+                        local stream_key = KEYS[1]
+                        local known_key = KEYS[2]
+                        local parked_key = KEYS[3]
+                        local queue_key = KEYS[4]
 
-            pipeline.set(known_task_key, when.timestamp())
+                        local task_key = ARGV[1]
+                        local when_timestamp = ARGV[2]
+                        local is_immediate = ARGV[3] == '1'
+                        local replace = ARGV[4] == '1'
 
-            if when <= datetime.now(timezone.utc):
-                pipeline.xadd(self.stream_key, message)  # type: ignore[arg-type]
-            else:
-                pipeline.hset(self.parked_task_key(key), mapping=message)  # type: ignore[arg-type]
-                pipeline.zadd(self.queue_key, {key: when.timestamp()})
+                        -- Extract message fields from ARGV[5] onwards
+                        local message = {}
+                        for i = 5, #ARGV, 2 do
+                            message[#message + 1] = ARGV[i]     -- field name
+                            message[#message + 1] = ARGV[i + 1] -- field value
+                        end
 
-    async def _cancel(self, pipeline: Pipeline, key: str) -> None:
-        pipeline.delete(self.known_task_key(key))
-        pipeline.delete(self.parked_task_key(key))
-        pipeline.zrem(self.queue_key, key)
+                        -- Handle replacement: cancel existing task if needed
+                        if replace then
+                            local existing_message_id = redis.call('HGET', known_key, 'stream_message_id')
+                            if existing_message_id then
+                                redis.call('XDEL', stream_key, existing_message_id)
+                            end
+                            redis.call('DEL', known_key, parked_key)
+                            redis.call('ZREM', queue_key, task_key)
+                        else
+                            -- Check if task already exists
+                            if redis.call('EXISTS', known_key) == 1 then
+                                return 'EXISTS'
+                            end
+                        end
+
+                        if is_immediate then
+                            -- Add to stream and store message ID for later cancellation
+                            local message_id = redis.call('XADD', stream_key, '*', unpack(message))
+                            redis.call('HSET', known_key, 'when', when_timestamp, 'stream_message_id', message_id)
+                            return message_id
+                        else
+                            -- Add to queue with task data in parked hash
+                            redis.call('HSET', known_key, 'when', when_timestamp)
+                            redis.call('HSET', parked_key, unpack(message))
+                            redis.call('ZADD', queue_key, when_timestamp, task_key)
+                            return 'QUEUED'
+                        end
+                        """
+                    ),
+                )
+            schedule_task = self._schedule_task_script
+
+            await schedule_task(
+                keys=[
+                    self.stream_key,
+                    known_task_key,
+                    self.parked_task_key(key),
+                    self.queue_key,
+                ],
+                args=[
+                    key,
+                    str(when.timestamp()),
+                    "1" if is_immediate else "0",
+                    "1" if replace else "0",
+                    *[
+                        item
+                        for field, value in message.items()
+                        for item in (field, value)
+                    ],
+                ],
+            )
+
+    async def _cancel(self, redis: Redis, key: str) -> None:
+        """Cancel a task atomically.
+
+        Handles cancellation regardless of task location:
+        - From the stream (using stored message ID)
+        - From the queue (scheduled tasks)
+        - Cleans up all associated metadata keys
+        """
+        if self._cancel_task_script is None:
+            self._cancel_task_script = cast(
+                _cancel_task,
+                redis.register_script(
+                    # KEYS: stream_key, known_key, parked_key, queue_key
+                    # ARGV: task_key
+                    """
+                    local stream_key = KEYS[1]
+                    local known_key = KEYS[2]
+                    local parked_key = KEYS[3]
+                    local queue_key = KEYS[4]
+                    local task_key = ARGV[1]
+
+                    -- Delete from stream if message ID exists
+                    local message_id = redis.call('HGET', known_key, 'stream_message_id')
+                    if message_id then
+                        redis.call('XDEL', stream_key, message_id)
+                    end
+
+                    -- Clean up all task-related keys
+                    redis.call('DEL', known_key, parked_key)
+                    redis.call('ZREM', queue_key, task_key)
+
+                    return 'OK'
+                    """
+                ),
+            )
+        cancel_task = self._cancel_task_script
+
+        # Execute the cancellation script
+        await cancel_task(
+            keys=[
+                self.stream_key,
+                self.known_task_key(key),
+                self.parked_task_key(key),
+                self.queue_key,
+            ],
+            args=[key],
+        )
 
     @property
     def strike_key(self) -> str:

--- a/src/docket/worker.py
+++ b/src/docket/worker.py
@@ -406,7 +406,7 @@ class Worker:
                             task[task_data[j]] = task_data[j+1]
                         end
 
-                        redis.call('XADD', KEYS[2], '*',
+                        local message_id = redis.call('XADD', KEYS[2], '*',
                             'key', task['key'],
                             'when', task['when'],
                             'function', task['function'],
@@ -414,6 +414,9 @@ class Worker:
                             'kwargs', task['kwargs'],
                             'attempt', task['attempt']
                         )
+                        -- Store the message ID in the known task key
+                        local known_key = ARGV[2] .. ":known:" .. key
+                        redis.call('HSET', known_key, 'stream_message_id', message_id)
                         redis.call('DEL', hash_key)
                         due_work = due_work + 1
                     end

--- a/tests/test_fundamentals.py
+++ b/tests/test_fundamentals.py
@@ -250,10 +250,10 @@ async def test_cancelling_future_task(
     the_task.assert_not_called()
 
 
-async def test_cancelling_current_task_not_supported(
+async def test_cancelling_immediate_task(
     docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
 ):
-    """docket does not allow cancelling a task that is schedule now"""
+    """docket can cancel a task that is scheduled immediately"""
 
     execution = await docket.add(the_task, now())("a", "b", c="c")
 
@@ -261,7 +261,28 @@ async def test_cancelling_current_task_not_supported(
 
     await worker.run_until_finished()
 
-    the_task.assert_awaited_once_with("a", "b", c="c")
+    the_task.assert_not_called()
+
+
+async def test_cancellation_is_idempotent(
+    docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
+):
+    """Test that canceling the same task twice doesn't error."""
+    key = f"test-task:{uuid4()}"
+
+    # Schedule a task
+    later = now() + timedelta(seconds=1)
+    await docket.add(the_task, later, key=key)("test")
+
+    # Cancel it twice - both should succeed without error
+    await docket.cancel(key)
+    await docket.cancel(key)  # Should be idempotent
+
+    # Run worker to ensure the task was actually cancelled
+    await worker.run_until_finished()
+
+    # Task should not have been executed since it was cancelled
+    the_task.assert_not_called()
 
 
 async def test_errors_are_logged(

--- a/tests/test_worker.py
+++ b/tests/test_worker.py
@@ -2,8 +2,9 @@ import asyncio
 import logging
 from contextlib import asynccontextmanager
 from datetime import datetime, timedelta, timezone
-from typing import AsyncGenerator
+from typing import AsyncGenerator, Callable
 from unittest.mock import AsyncMock, patch
+from uuid import uuid4
 
 import pytest
 from redis.asyncio import Redis
@@ -1427,3 +1428,79 @@ async def test_finally_block_releases_concurrency_on_success(docket: Docket):
 
     # If both tasks completed, the finally block successfully released slots
     assert task_completed
+
+
+async def test_replacement_race_condition_stream_tasks(
+    docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
+):
+    """Test that replace() properly cancels tasks already in the stream.
+
+    This reproduces the race condition where:
+    1. Task is scheduled for immediate execution
+    2. Scheduler moves it to stream
+    3. replace() tries to cancel but only checks queue/hash, not stream
+    4. Both original and replacement tasks execute
+    """
+    key = f"my-cool-task:{uuid4()}"
+
+    # Schedule a task immediately (will be moved to stream quickly)
+    await docket.add(the_task, now(), key=key)("a", "b", c="c")
+
+    # Let the scheduler move the task to the stream
+    # The scheduler runs every 250ms by default
+    await asyncio.sleep(0.3)
+
+    # Now replace the task - this should cancel the one in the stream
+    later = now() + timedelta(milliseconds=100)
+    await docket.replace(the_task, later, key=key)("b", "c", c="d")
+
+    # Run the worker to completion
+    await worker.run_until_finished()
+
+    # Should only execute the replacement task, not both
+    the_task.assert_awaited_once_with("b", "c", c="d")
+    assert the_task.await_count == 1, (
+        f"Task was called {the_task.await_count} times, expected 1"
+    )
+
+
+async def test_replace_task_in_queue_before_stream(
+    docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
+):
+    """Test that replace() works correctly when task is still in queue."""
+    key = f"my-cool-task:{uuid4()}"
+
+    # Schedule a task slightly in the future (stays in queue)
+    soon = now() + timedelta(seconds=1)
+    await docket.add(the_task, soon, key=key)("a", "b", c="c")
+
+    # Replace immediately (before scheduler can move it)
+    later = now() + timedelta(milliseconds=100)
+    await docket.replace(the_task, later, key=key)("b", "c", c="d")
+
+    await worker.run_until_finished()
+
+    # Should only execute the replacement
+    the_task.assert_awaited_once_with("b", "c", c="d")
+    assert the_task.await_count == 1
+
+
+async def test_rapid_replace_operations(
+    docket: Docket, worker: Worker, the_task: AsyncMock, now: Callable[[], datetime]
+):
+    """Test multiple rapid replace operations."""
+    key = f"my-cool-task:{uuid4()}"
+
+    # Schedule initial task
+    await docket.add(the_task, now(), key=key)("a", "b", c="c")
+
+    # Rapid replacements
+    for i in range(5):
+        when = now() + timedelta(milliseconds=50 + i * 10)
+        await docket.replace(the_task, when, key=key)(f"arg{i}", b=f"b{i}")
+
+    await worker.run_until_finished()
+
+    # Should only execute the last replacement
+    the_task.assert_awaited_once_with("arg4", b="b4")
+    assert the_task.await_count == 1

